rebalancing/rebalancing.py

```diff
@@ -0,0 +1,311 @@
+        """
+        :param _value: sICX amount to set.
+        Sets the sICX amount to receive by rebalancing contract.
+        """
+        self._sicx_receivable.set(_value)
+
+    @external
+    @only_governance
+    def setBnusdReceivable(self, _value: int) -> None:
+        """
+        :param _value: bnUSD amount to set.
+        Sets the bnUSD amount to receive by rebalancing contract.
+        """
+        self._bnusd_receivable.set(_value)
+
+    @external(readonly=True)
+    def getSicxReceivable(self) -> int:
+        """
+        Returns the sICX amount to receive by rebalancing contract.
+        """
+        return self._sicx_receivable.get()
+
+    @external(readonly=True)
+    def getBnusdReceivable(self) -> int:
+        """
+        Returns the bnUSD amount to receive by rebalancing contract.
+        """
+        return self._bnusd_receivable.get()
+
+    @external(readonly=True)
+    def getRebalancingStatus(self) -> list:
+        """
+        Checks the Rebalancing status of the pool.
+        """
+        rebalancing_direction = 0
+        self.bnUSD_score = self.create_interface_score(self._bnUSD.get(), bnUSDTokenInterface)
+        self.dex_score = self.create_interface_score(self._dex.get(), dexTokenInterface)
+        self.loans_score = self.create_interface_score(self._loans.get(), loansTokenInterface)
+        price = self.bnUSD_score.lastPriceInLoop()
+        pool_price_dex = self.dex_score.getPriceByName("sICX/bnUSD")
+        difference = price - (10 ** 36 // pool_price_dex)
+        if price < (10 ** 36 // pool_price_dex):
+            difference = (10 ** 36 // pool_price_dex) - price
+            rebalancing_direction = 1
+        change_in_percent = self._change_in_percent(price, difference)
+        if change_in_percent > self._price_threshold.get():
+            if rebalancing_direction == 1:
+                return [True, self._calculate_tokens_to_retire(self.dex_score, 1), "bnUSD"]
+            return [True, self._calculate_tokens_to_retire(self.dex_score, 0), "sICX"]
+        else:
+            return [False, 0]
+
+    def _change_in_percent(self, price: int, difference: int) -> int:
+        return (difference * 10 ** 18 // price) * 100
+
+    @external
+    def rebalance(self) -> None:
+        """
+           Calls the retireRedeem function of loans and retire the value of bnUSD.
+           Rebalances only if the rate of change in dex pool price and oracle price is greater than the threshold set.
+        """
+        self.sICX_score = self.create_interface_score(self._sicx.get(), sICXTokenInterface)
+        self.bnUSD_score = self.create_interface_score(self._bnUSD.get(), bnUSDTokenInterface)
+        rebalancing_status = self.getRebalancingStatus()
+        if rebalancing_status[0]:
+            if rebalancing_status[2] == "sICX":
+                sicx_to_retire = rebalancing_status[1]
+                sicx_in_contract = self.sICX_score.balanceOf(self.address)
```

Comment:
Right now there is no control over the tokenFallback so anyone could change the sicx_to_retire amount by adding more sicx. Probably best to make this a configurable parameter. Could retire the lesser of the parameter amount or the SCORE balance.

Reply:
Do you mean setting a parameter by owner that sets 1000 value for now?

Reply:
Didn't quite get you.
But In rebalancing contract token fallback, we will revert back all the sICX sending transaction if it is not from the loans contract or owner. In that case the sicx_to_retire won't be changed.